### PR TITLE
Enable IPv6 URIs in reindex from remote

### PR DESCRIPTION
**Pull request by hub-cap**
Reindex from remote was using a custom regex to dermine what URIs were
valid. This commit removes the custom regex and uses the java.net.URI
class instead, allowing IPv6 support without changing the existing
validation around a URI in reindex from remote.

### CODE REVIEW

**Review by ywelsch — CHANGES_REQUESTED** [review]
Would it be possible to have a test that runs some actual reindex operations against `::1` if available?

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java [hunk 1]
```diff
@@ -136,15 +135,27 @@ static RemoteInfo buildRemoteInfo(Map<String, Object> source) throws IOException
...
-        String host = hostMatcher.group("host");
-        String pathPrefix = hostMatcher.group("pathPrefix");
-        int port = Integer.parseInt(hostMatcher.group("port"));
+
+        String scheme = uri.getScheme();
+        String host = uri.getHost();
+        int port = uri.getPort();
+
+        String pathPrefix = null;
+        if (uri.getPath().isEmpty() == false) {
```
[the remark this answers is not in the document]
Comment (hub-cap) [reply modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java]: there is an existing test that actually wants it to be null, so i wanted to keep that functionality. It used to return null in the regex, and now it was returning an empty string, so i kept the old functionality.

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java [hunk 2]
```diff
@@ -136,15 +135,27 @@ static RemoteInfo buildRemoteInfo(Map<String, Object> source) throws IOException
...
-                + hostInRequest + "]");
+                + hostInRequest + "]", ex);
         }
-        String scheme = hostMatcher.group("scheme");
-        String host = hostMatcher.group("host");
-        String pathPrefix = hostMatcher.group("pathPrefix");
-        int port = Integer.parseInt(hostMatcher.group("port"));
+
+        String scheme = uri.getScheme();
+        String host = uri.getHost();
```
[the remark this answers is not in the document]
Comment (hub-cap) [reply modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java]: Ill add a null host check in RestReindexActionTests, but the URI wont parse if its null, so I dont have to add any code here.

https://github.com/elastic/elasticsearch/blob/master/modules/reindex/src/test/java/org/elasticsearch/index/reindex/RestReindexActionTests.java#L79-L83

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java [hunk 3]
```diff
@@ -136,15 +135,27 @@ static RemoteInfo buildRemoteInfo(Map<String, Object> source) throws IOException
...
             throw new IllegalArgumentException("[host] must be of the form [scheme]://[host]:[port](/[pathPrefix])? but was ["
-                + hostInRequest + "]");
+                + hostInRequest + "]", ex);
         }
-        String scheme = hostMatcher.group("scheme");
-        String host = hostMatcher.group("host");
-        String pathPrefix = hostMatcher.group("pathPrefix");
-        int port = Integer.parseInt(hostMatcher.group("port"));
+
+        String scheme = uri.getScheme();
```
[the remark this answers is not in the document]
Comment (hub-cap) [reply modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java]: This is already handled in this test https://github.com/elastic/elasticsearch/blob/master/modules/reindex/src/test/java/org/elasticsearch/index/reindex/RestReindexActionTests.java#L79-L83

File: modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (hub-cap) [reply modules/reindex/src/main/java/org/elasticsearch/index/reindex/RestReindexAction.java]: Added some tests in https://github.com/elastic/elasticsearch/pull/36874/commits/4229f604c127730f58c39cac378e432686cae9b2

**Comment by hub-cap** [comment]
WRT testing `[::1]`, I have a branch that I will be PR'ing once this is merged taht does precisely this in the HLRC.

https://github.com/hub-cap/elasticsearch/commit/d8bfbc870d50c8098bc63c9c4e6803aad0b9088b

**Review by ywelsch — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**